### PR TITLE
`@purplet/utils`

### LINKED ISSUE #51: Purplet depends on `@discordjs/rest`'s `cdn` object which we didn't implement
We didn't implement this feature because it's not rest related, it's just url generating. We can create some utilities at some point. Will probably be closed during dev of purplet structures

### PR DESCRIPTION
**Pull request by Clembs**
This would be a new package, of random utilities like constants, functions, regex, etc for Discord API driven development.

Closes #51

### CODE REVIEW

**Comment by Clembs** [comment]
@paperdave tell me if you like that first draft commit so i can take your feedback into account when finishing the end product

**Comment by paperdave** [comment]
good work so far. my notes:

- can you copy `packages/purplet/src/utils/json.ts` into it. i was going to ask this for a few other files but after checking, i think those utils (like `stripUndefined` from the gateway client) might just go in my personal utils package; this package is for discord api related stuff.

- in `src/format/mentions.ts`, please refactor to use `export function ...` syntax (i am aware that arrow functions would be minified a bit smaller)

- i'd like `userMention`, `channelMention`, and `roleMention` to all be usable with just the `id` of the target. like `userMention('327690719085068289')`. right now the types for these functions require the full data present. two implementation ideas for this
  - only accept strings, require `userMention(userObj.id)`
  - accept `string | { id: string }` allowing both, example implementation could look like
  ``` ts
  export type IdObject = { id: string };
  export function mentionUser(x: IdObject | string) {
    return `<@${(x as IdObject).id ?? x}>`;
  }
  ```

- similar thing for `slashCommandMention` and `getUserDisplayAvatarURL` but the solution is a bit more complicated and i'm not entirely sure on it. 

- i do think for the current image link formatter, format and size could be an object, which will make a bit more sense when theres the whole list of image link formatters. maybe the way this would work for `getUserDisplayAvatarURL` is `getUserDisplayAvatarURL(user.id, user.avatar, { format: 'jpeg' })`

- CdnImageAllowedFormats, should just be named `CDNImageFormat`, and the size one being `CDNImageSize`. When i name those enums i usually make them a singular word.

- `customEmojiMention` is fine since the object passed is just the metadata needed for the formatter. maybe it could be renamed to just `emojiMention` since all other type names for custom emojis just call them _emojis_. then once this is done probably end up removing the hilarious `standardEmojiMention`, while it's a great joke it might cause confusion when people dive into the docs (issue for another time).

- something we could add later is `dateToSnowflake`. dont worry about this for now (issue for another time).

- what do you think about moving the builder classes we have into this package? or do you think we should make that separate? (either way, do not move them yet)

**Comment by paperdave** [comment]
for `getUserDisplayAvatarURL`, how about we go with two functions, a dumb one that just makes the url. and the another that does smart decision making version with `Pick<APIUser, 'id'> & Partial<APIUser, 'discriminator' | 'avatar'>` as the input type. allowing people with exact data (such as from a database with just `id` + `avatar` data), vs people with full/partial user objects. we definetly want to have functions to avoid users to hardcode the full urls.

how these are named. no idea

**Comment by paperdave** [comment]
i've added a `resolveUserAvatarURL` function which replaces `getUserDisplayAvatarURL`, as well as a `resolveMemberAvatarURL` function. these take api structures, where as the acompanying `format**URL` functions in `cdn.ts` are just string creator functions and have little logic.

**Comment by Clembs** [comment]
Tested the package as a whole, and merged component builders from the `@purplet/builders`, which we should be able to safely retire. This is like 95% completed (besides what we could add later), and I think we could merge this to main.

File: packages/utils/src/format/mentions.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  ShortTime = 't',
+  LongTime = 'T',
+  ShortDate = 'd',
+  LongDate = 'D',
+  ShortDateTime = 'f',
+  LongDateTime = 'F',
+  Relative = 'R',
+}
+
+export function timestampMention(date: Date | number, format?: TimestampFormats) {
```
Comment (paperdave) [inline packages/utils/src/format/mentions.ts:61]: didn't you want to have dayjs support here, or are those objects valid as a `Date`?

**Review by paperdave — APPROVED** [review]
good work. do look at my comment on the timestampMention, but it can be looked at later.

File: packages/utils/src/format/mentions.ts (hunk 1 above)
Comment (Clembs) [reply packages/utils/src/format/mentions.ts:61]: that can come later, for now to use dayjs you can `dayjs().toDate()`.

**Pull request merged**